proposals/0000-extract-runtime.md

```diff
@@ -0,0 +1,419 @@
+without impacting `react-native` library.
+
+This also makes the updating process easier.
+
+For matching the `react-native-runtime` and `react-native` versions, we could use:
+
+ - peer dependency in `react-native-core`
+ - `package.json` `engine` field
+
+### Codegen
+
+Because the TurboModule bridging would live in `react-native-runtime`, the codegen would need to be updated.
+
+### Old architecture
+
+As Old architecture is deprecated in favor of new architecture. The bridge can be kept in `react-native`
+for transition period.
+
+### Transition period
+
+There are several approaches we can perform to allow transitioning to new approach:
+
+#### C++ Code
+
+For transition period, we can re-export the symbols from `react-native-runtime` in `react-native` headers.
+A new C++ compile-time flag can be added to allow branching in the code depending on whenever the runtime is available or not.
+
+```cpp
+#ifdef RN_HAS_RUNTIME
+#include <react-native-runtime.h>
+#else
+#include <react-native.h>
+#endif
+```
+
+or alternatively using [`__has_include` macro](https://clang.llvm.org/docs/LanguageExtensions.html#has-include):
+
+```cpp
+#if __has_include(<react-native-runtime.h>)
+#include <react-native-runtime.h>
+#endif
+```
+
+#### Application code
+
+For end-user applications we can reduce the impact and make `react-native` link the `react-native-runtime` library 
+(either statically or dynamically) by default. User could opt out of this feature and then link the runtime
+library manually.
+
+This conditional dependency would need to live in both `react-native` and (reversed) in user project during the
+transition.
+
+Conditional dependencies are supported by all build systems:
+
+ - [Gradle](https://www.baeldung.com/gradle-conditional-dependencies)
+ - CocoaPods
+
+### Potential `react-native-core` extraction
+
+Mentioned as a possibility during Core Contributors Summit 2024, [extracting platforms from `react-native`](https://github.com/react-native-community/discussions-and-proposals/pull/49), and therefore
+creating a `react-native-core` layer from would fit perfectly with the suggested approach.
+
+`react-native-core` would then depend on `react-native-runtime` implementing generic UI code, exposing ShadowNode API
+allowing other platforms to depend on it.
+
+Such separation of layers would make it more analogous to how React Web is structured right now:
+
+![Layers comparison](./assets/0000-rn-diagram-matching-layers.png)
+
+### Allowing more OOT composability
+
+If at any point a new platform needs to make alternations in the runtime it still can
+use existing `react-native` package for the UI. The same goes for the inverse if a platform,
+for any reason, needs to have its own runtime.
+
+## Drawbacks
+
+ - Extracting API surface layer is always a big task cost and time-wise
+ - We have to migrate to this approach affecting external libraries and applications.
```

Comment:
Do you mean that "packages wanting to leverage the possibilities enabled by RNR would need to migrate to a completely new API" or something else? We can't break existing libraries by adding this new layer, it should be opaque from a library point of view

Reply:
Updated the wording